src/VkLayer_GF_amber_scoop/src/buffer_copy.cc

```diff
@@ -0,0 +1,208 @@
+
+namespace gf_layers::amber_scoop_layer {
+
+BufferCopy::BufferCopy(
+    DeviceData* device_data, const VkBuffer& buffer, VkDeviceSize buffer_size,
+    VkQueue queue, VkCommandPool command_pool,
+    const std::vector<const CmdPipelineBarrier*>& pipeline_barriers)
+    : device_data_(device_data) {
+  VkDevice device = device_data_->device;
+
+  // Create a buffer where the data will be copied to.
+  {
+    VkBufferCreateInfo create_info = {};
+    create_info.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO;
+    create_info.size = buffer_size;
+    create_info.usage = VK_BUFFER_USAGE_TRANSFER_DST_BIT;
+    create_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+
+    RequireSuccess(device_data_->vkCreateBuffer(device, &create_info, nullptr,
+                                                &buffer_copy_),
+                   "Failed to create buffer for the copied data.");
+  }
+
+  VkMemoryRequirements memory_requirements;
+  device_data_->vkGetBufferMemoryRequirements(device, buffer_copy_,
+                                              &memory_requirements);
+  {
+    VkMemoryAllocateInfo vkMemoryAllocateInfo = {};
+    vkMemoryAllocateInfo.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+    vkMemoryAllocateInfo.allocationSize = memory_requirements.size;
+    vkMemoryAllocateInfo.memoryTypeIndex =
+        FindMemoryType(memory_requirements.memoryTypeBits,
+                       VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
+    RequireSuccess(
+        device_data_->vkAllocateMemory(device, &vkMemoryAllocateInfo, nullptr,
+                                       &buffer_copy_memory_),
+        "Failed to allocate memory for buffer copy.");
+  }
+  RequireSuccess(device_data_->vkBindBufferMemory(device, buffer_copy_,
+                                                  buffer_copy_memory_, 0),
+                 "Failed binding memory for buffer copy.");
+
+  // Create command buffer
+  VkCommandBuffer command_buffer = nullptr;
+  VkCommandBufferAllocateInfo command_buffer_allocate_info = {};
+  command_buffer_allocate_info.sType =
+      VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+  command_buffer_allocate_info.commandPool = command_pool;
+  command_buffer_allocate_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+  command_buffer_allocate_info.commandBufferCount = 1;
+  RequireSuccess(device_data_->vkAllocateCommandBuffers(
+                     device, &command_buffer_allocate_info, &command_buffer),
+                 "Failed to allocate command buffers.");
+
+  // Record a command buffer for the copy operation.
+  VkCommandBufferBeginInfo begin_info = {};
+  begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+
+  RequireSuccess(
+      device_data_->vkBeginCommandBuffer(command_buffer, &begin_info),
+      "Failed to begin command buffer.");
+
+  // Copy all global and buffer memory barriers. Set
+  // VK_ACCESS_TRANSFER_READ_BIT for copied (buffer)memory barriers, so we can
+  // execute a copy operation.
```

Comment:
Yeah, I guess it's quite unlikely to see any barriers for vertex or index buffers, but it's still possible to have a case where a compute shader manipulates the vertex data and that might need some barrier(s). Quite unlikely though. There might be also some queue family ownership transfers, but that would be very unlikely too. I guess it would be indeed a lot easier and safer to use some predefined barriers.